server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/ACLUsernameChangeTaskStep.java

```diff
@@ -0,0 +1,95 @@
+import org.reactivestreams.Publisher;
+
+import com.github.fge.lambdas.Throwing;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class ACLUsernameChangeTaskStep implements UsernameChangeTaskStep {
+    private final MailboxManager mailboxManager;
+    private final SubscriptionManager subscriptionManager;
+
+    @Inject
+    public ACLUsernameChangeTaskStep(MailboxManager mailboxManager, SubscriptionManager subscriptionManager) {
+        this.mailboxManager = mailboxManager;
+        this.subscriptionManager = subscriptionManager;
+    }
+
+    @Override
+    public StepName name() {
+        return new StepName("ACLUsernameChangeTaskStep");
+    }
+
+    @Override
+    public int priority() {
+        return 3;
+    }
+
+    @Override
+    public Publisher<Void> changeUsername(Username oldUsername, Username newUsername) {
+        MailboxSession oldSession = mailboxManager.createSystemSession(oldUsername);
+        MailboxSession newSession = mailboxManager.createSystemSession(newUsername);
+        return mailboxManager.search(MailboxQuery.builder().matchesAllMailboxNames().build(), oldSession)
+            .filter(mailbox -> !mailbox.getPath().getUser().equals(oldUsername))
+            .concatMap(mailbox -> migrateACLs(oldUsername, newUsername, mailbox))
+            .then(updateSubscriptionsOnDeletedMailboxes(oldUsername, oldSession, newSession));
+    }
+
+    private Mono<Void> updateSubscriptionsOnDeletedMailboxes(Username oldUsername, MailboxSession oldSession, MailboxSession newSession) {
+        try {
+            return Flux.from(subscriptionManager.subscriptionsReactive(oldSession))
+                .filter(subscription -> !subscription.getUser().equals(oldUsername))
+                .concatMap(subscription -> Mono.from(subscriptionManager.subscribeReactive(subscription, newSession))
+                    .then(Mono.from(subscriptionManager.unsubscribeReactive(subscription, oldSession))))
+                .then();
+        } catch (SubscriptionException e) {
+            return Mono.error(e);
+        }
+    }
+
+    private Publisher<? extends Void> migrateACLs(Username oldUsername, Username newUsername, MailboxMetaData mailbox) {
+        MailboxSession ownerSession = mailboxManager.createSystemSession(mailbox.getPath().getUser());
+        MailboxACL.Rfc4314Rights rights = Optional.ofNullable(mailbox.getMailbox().getACL().getEntries().get(MailboxACL.EntryKey.createUserEntryKey(oldUsername)))
+            .orElse(MailboxACL.NO_RIGHTS);
+
+        return Mono.fromRunnable(Throwing.runnable(() -> mailboxManager.applyRightsCommand(mailbox.getId(), MailboxACL.command().rights(rights).forUser(newUsername).asAddition(), ownerSession)))
+            .then(Mono.fromRunnable(Throwing.runnable(() -> mailboxManager.applyRightsCommand(mailbox.getId(), MailboxACL.command().rights(rights).forUser(oldUsername).asRemoval(), ownerSession))))
+            .subscribeOn(ReactorUtils.BLOCKING_CALL_WRAPPER)
```

Comment:
Do we need to subscribe to `BLOCKING_CALL_WRAPPER` at step level while we already wrap the blocking scheduler at the task level?

Reply:
I think the step should not do assumptions on where it would be run and assume the worst...